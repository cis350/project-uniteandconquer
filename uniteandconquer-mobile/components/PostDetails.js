import {
  React, useState, useEffect,
} from 'react';
import {
  StyleSheet, View, ScrollView, Text, Button, TouchableOpacity, TextInput,
} from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';
import { leaveGroup } from '../modules/PostDB';

const PostDB = require('../modules/PostDB');

// styling ---------

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#FAE9C7',
    height: '100%',
    width: '100%',
  },
});

const userStyles = StyleSheet.create({
  container: {
    alignItems: 'stretch',
    justifyContent: 'center',
    margin: 10,
  },
  text: {
    textAlign: 'right',
    margin: 4,
    fontSize: 18,
    marginBottom: 10,
  },
  button: {
    alignSelf: 'flex-end',
  },
});

const postDetailStyles = StyleSheet.create({
  container: {
    borderWidth: 8,
    borderColor: '#FFD9A0',
    height: '90%',
  },
  tags: {
    backgroundColor: '#FFCB7D',
    width: '35%',
    marginRight: 10,
    alignItems: 'center',
    justifyContent: 'center',
  },
  upper: {
    display: 'flex',
    flexDirection: 'row',
    marginBottom: 10,
  },
  detailsHeader: {
    fontSize: 25,
    fontWeight: 'bold',
  },
  groupHeader: {
    fontSize: 25,
    fontWeight: 'bold',
  },
  tagsHeader: {
    fontSize: 15,
    fontWeight: 'bold',
  },
  tag: {
    borderWidth: 5,
    borderColor: '#FFFFFF',
    borderRadius: 15,
    padding: 3,
    backgroundColor: '#FFAB2D',
    overflow: 'hidden',
    margin: 4,
  },
  details: {
    marginBottom: 15,
    marginTop: 7,
    fontSize: 15,
  },
  groupDetailsContainer: {
    marginLeft: 8,
  },
  groupUser: {
    display: 'flex',
    flexDirection: 'row',
    marginBottom: 10,
    width: '40%',
  },
  groupUserName: {
    fontSize: 20,
    marginRight: 20,
  },
  groupUserQuantity: {
    fontSize: 20,
    marginRight: 20,
  },
  groupUserIcon: {
    marginRight: 20,
  },
  LeftButton: {
    backgroundColor: '#FFCB7D',
    borderRadius: 6,
    padding: 3,
    marginRight: 5,
  },
  RightButton: {
    backgroundColor: '#C6C6C6',
    borderRadius: 6,
    padding: 3,
    marginLeft: 5,
  },
  buttons: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    position: 'relative',
    bottom: -50,
    marginBottom: 100,
  },
  upperBox: {
    padding: 5,
    backgroundColor: '#FFD9A0',
    width: '40%',
    textAlign: 'center',
    fontSize: 20,
    fontWeight: 'bold',
  },
  groupUserContent: {
    flexDirection: 'row',
    alignItems: 'center',
    margin: 10,
  },
});

// app content --------

export default function PostDetails({ navigation }) {
  // event handlers --------
  const [errorMessage, setErrorMessage] = useState(null);
  const [join, setJoin] = useState(false);
  const [members, setMembers] = useState([]);
  const [ownerID, setOwnerID] = useState('');
  const [quantity, setQuantity] = useState(0);
  // const [memberId, setMemberId] = useState('');
  const postId = '1';// need to be change.
  const userId = '5087901e810c109679e860ea';

  const leavePressed = () => {
    // const userId = ''; // dummy, need to get from session once log in routing is set up
    // const postId = ''; // dummy, need to get from current view
    leaveGroup(userId, postId, (success, err) => {
      setErrorMessage(err); // trigger re-rendering; will display err if err is not null
    });
    setJoin(false);
  };
  const handleJoin = () => {
    const userID = 'bababababa';
    const postID = 0;
    // const quantity = 2;
    PostDB.joinGroup(userID, postID, quantity, (success, err) => {
      if (success) {
        setJoin(true);
        PostDB.getPostMembers(postId, (success2, memberList, err2) => {
          if (success2) {
            setMembers(memberList);
          } else {
            console.log(err2);
          }
        });
      } else { setErrorMessage(err); }
    });
  };

  const handleJoinLeave = () => {
    if (join) {
      leavePressed();
    } else {
      handleJoin();
    }
  };

  useEffect(() => {
    PostDB.getPostMembers(postId, (success, memberList, err) => {
      if (success) {
        console.log(memberList);
        setMembers(memberList);
        if (memberList.some((e) => e.id === userId)) {
          setJoin(true);
        }
      } else {
        console.log(err);
      }
      PostDB.getOwner(postId, (success2, owner, err2) => {
        if (success2) {
          setOwnerID(owner);
        } else {
          console.log(err2);
        }
      });
    });
  }, []);
  const handleKick = (memberID) => {
    PostDB.kickMembers(memberID, postId, (success, err) => {
      if (success) {
        console.log(members.filter((item) => item.id !== memberID));
        setMembers(members.filter((item) => item.id !== memberID));
      } else {
        console.log(err);
      }
    });
  };
  function kickButton(memberID) {
    console.log(ownerID === userId);
    if (ownerID === userId && ownerID !== memberID) {
      return (
        <View style={postDetailStyles.LeftButton}>
          <Button color="#000" title="Kick" onPress={() => handleKick(memberID)} />
        </View>
      );
    }
    return null;
  }

  // views ---------

  return (
    <ScrollView style={styles.container}>
      <View style={userStyles.container}>
        <TouchableOpacity
          style={{ alignSelf: 'flex-end' }}
          onPress={() => navigation.navigate('UserProfile')}
        >
          <Icon name="user" size={28} style={userStyles.icon} />
        </TouchableOpacity>
        <Text style={userStyles.text}>
          Hello, guest!
        </Text>
        <Text style={userStyles.button}>
          <Button
            color="#000"
            title="My Chats"
          />
        </Text>
        <Text style={userStyles.button}>
          <Button
            color="#000"
            title="Log In"
            onPress={() => navigation.navigate('LogIn')}
          />
        </Text>
      </View>
      {errorMessage && (
        <Text>{errorMessage}</Text>
      )}
      <Text style={postDetailStyles.upperBox}>Item Name</Text>
      <View style={postDetailStyles.container}>
        <View style={postDetailStyles.upper}>
          <View style={postDetailStyles.tags}>
            <Text style={postDetailStyles.tagsHeader}>Tags</Text>
            <View>
              <Text style={postDetailStyles.tag}>Tag1</Text>
              <Text style={postDetailStyles.tag}>Tag2</Text>
              <Text style={postDetailStyles.tag}>Tag3</Text>
              <Text style={postDetailStyles.tag}>Tag3</Text>
              <Text style={postDetailStyles.tag}>Tag3</Text>
            </View>
          </View>
          <View style={{ flexShrink: 1 }}>
            <Text style={postDetailStyles.detailsHeader}>Item Details</Text>
            <View>
              <Text style={postDetailStyles.details}>Target Quantity: 4/5</Text>
              <Text style={postDetailStyles.details}>Price/Item: $10.00</Text>
              <Text style={postDetailStyles.details}>Item Link example.com/item</Text>
              <Text style={postDetailStyles.details}>
                Description: Beautiful fairy lights to use for your dorm
              </Text>
            </View>
          </View>
        </View>
        <View style={postDetailStyles.groupDetailsContainer}>
          <Text style={postDetailStyles.groupHeader}>Group Details</Text>
          <Text style={postDetailStyles.details}>
            Group size:
            {' '}
            { members.length}
          </Text>
          <View>
            <View style={postDetailStyles.groupUser}>
              <View>
                {members.map((item) => (
                  <View key={item.name} style={postDetailStyles.groupUserContent}>
                    <View>
                      <Icon
                        style={postDetailStyles.groupUserIcon}
                        name="user"
                        size={40}
                      />
                    </View>
                    <View>
                      <Text
                        style={postDetailStyles.groupUserName}
                      >
                        {item.name}
                      </Text>
                    </View>
                    <View>
                      <Text
                        style={postDetailStyles.groupUserQuantity}
                      >
                        Quantity:
                        {' '}
                        {item.quantity}
                      </Text>
                    </View>
                    {kickButton(item.id)}
                  </View>
                ))}
              </View>
            </View>
          </View>
          <View>
            <TextInput
              placeholder="quantity"
              onChangeText={setQuantity}
              value={quantity}
            />
            <View style={postDetailStyles.buttons}>
              <View style={postDetailStyles.LeftButton}><Button color="#000" title={join ? 'Leave' : 'Join'} onPress={handleJoinLeave} /></View>
              <View style={postDetailStyles.LeftButton}><Button color="#000" title="Comment" onPress={() => navigation.navigate('Comment')} /></View>
              <View style={postDetailStyles.RightButton}><Button color="#000" title="Back" onPress={() => navigation.navigate('Home')} /></View>

            </View>
          </View>
        </View>
      </View>
    </ScrollView>
  );
}
