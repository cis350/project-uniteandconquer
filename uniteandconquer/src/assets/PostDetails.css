.post-details-page {
  height: 100%;
  display: grid;
  grid-template-columns: 0.2fr 1fr;
}

.desired-quantity {
  max-width: 330px;
  margin-bottom: 10px;
}

.item-name {
  font-size: 25px;
}

.item-details {
  margin-right: 25px;
}

.post-details-container {
  margin-left: 10%;
}

.post-details-content {
  display: grid;
  grid-template-columns: 0.5fr 0.5fr;
}

ul {
  padding: 0;
}

.far.fa-user-circle {
  transform: translateY(-15%);
}

.group-user {
  justify-content: center;
  align-items: center;

  border: 5px solid #ffd9a0;
  margin-bottom: 20px;
  padding: 10px;
  display: flex;
  flex-direction: row;
  width: 50%;
}

.name {
  display: flex;
  flex-direction: row;
  margin-right: 30px;
}

li {
  display: grid;
  grid-template-columns: 0.3fr 0.1fr;
  width: 50%;
  font-size: 19px;
  margin: 15px;
}

.size {
  max-width: 400px;
  font-size: 19px;
  margin: 5px;
}

.group-members {
  text-align: left;
}

.item-detail {
  max-width: 400px;
  font-size: 19px;
  margin: 15px;
}

.post-detail-tags {
  margin-top: 10px;
  text-align: left;
}

.tags-container {
  display: grid;
  grid-template-columns: 0.25fr 0.25fr 0.25fr 0.25fr;
  width: 60%;
  margin-top: 10px;
  margin-bottom: 30px;
}

.tags-label {
  font-weight: bold;
}

.buttons-container {
  margin: auto;
  width: 50%;
  margin-bottom: 40px;
}

.horizontal-buttons {
  display: flex;
  flex-direction: row;
  margin: auto;
}

.create-button {
  border: none;
  background: #ffcb7d;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  border-radius: 5px;
  padding: 15px;
  width: 150px;

  font-style: italic;
  font-weight: bold;
  font-size: 20px;
  margin: 5px;
}

.cancel-button {
  border: none;
  background: #ffcb7d;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  border-radius: 5px;
  padding: 15px;
  width: 150px;

  font-style: italic;
  font-weight: bold;
  font-size: 20px;
  margin: 5px;
}


.delete-button {
  border: none;
  background: #ffcb7d;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  border-radius: 5px;
  padding: 15px;
  width: 150px;

  font-style: italic;
  font-weight: bold;
  font-size: 20px;
  margin: 5px;
  margin-left: 25%;
}

.comment-section {
  padding: 15px;
}

.row-setting {
  display: flex;
}

.cross-sign-wrapper {
  margin-top: 2%;
  margin-left: -10%;
}

.cross-sign {
  background: #ffcb7d;
}