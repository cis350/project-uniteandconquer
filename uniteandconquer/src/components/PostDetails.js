/* eslint-disable no-unused-vars */
import React, { useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import Sidebar from './Sidebar';
import Comment from './Comment';
import '../assets/PostDetails.css';

const PostDB = require('../modules/PostDB');

// In the post details page, one can choose to join the group
// if successful, redirect the user to the post which the user joins in
function PostDetails() {
  const [desiredQuantity, setDesiredQuantity] = useState(0);
  const [validQuantity, setValidQuantity] = useState(false);
  const navigate = useNavigate();
  const myStorage = window.sessionStorage;
  const [postDetails, setPostDetails] = useState(null);

  useEffect(() => {
    PostDB.getPost('', (success, details) => {
      if (success) {
        setPostDetails(details);
      }
    });
  }, []);
  const joinGroup = () => {
    const userID = myStorage.getItem('UserID');
    const postID = '';
    const quantity = 2;
    PostDB.joinGroup(userID, postID, quantity, (success, err) => {
      if (success) {
        navigate(`/post-details/${postID}`);
      } else {
        console.log(err);
      }
    });
  };

  const back = () => {
    navigate('/');
  };

  const quantityValidation = () => {
    const num = /^\d+$/;

    if (desiredQuantity) {
      if (desiredQuantity.match(num)) {
        setValidQuantity(true);
      } else {
        setValidQuantity(false);
      }
    } else {
      setValidQuantity(false);
    }
  };

  useEffect(() => {
    quantityValidation();
  });

  if (postDetails) {
    return (
      <div className="post-details-page">
        <Sidebar />
        <div className="post-details-container">
          <div className="item-name">
            <h1>{postDetails.itemName}</h1>
          </div>
          <div className="post-details-content">
            <div className="item-details">
              <h3>Item Details</h3>
              <div className="item-detail">
                Target Quantity:
                {' '}
                {postDetails.itemNumTarget}
              </div>

              <div className="item-detail">
                Current Quantity:
                {' '}
                {postDetails.itemNumCurrent}
              </div>

              <div className="item-detail">
                Price/Item: $
                {postDetails.pricePerItem}
              </div>

              <div className="item-detail">
                Item Link:
                {' '}
                <a href={postDetails.itemURL}>Click here to go the item page</a>
              </div>

              <div className="item-detail">
                Description:
                {' '}
                {postDetails.itemDescription}
              </div>

            </div>
            <div className="group-details">
              <h3>Group Details</h3>
              <div className="size">
                Group size:
                {' '}
                {postDetails.group.length}
              </div>
              <div className="group-member">
                <ul>
                  {postDetails.group.map((user) => (
                    <div className="group-user" key={user.firstName}>
                      <div className="name">
                        {' '}
                        <i className="far fa-user-circle" />
                        {' '}
                        {user.firstName}
                      </div>
                      <div className="quantity">
                        Quantity:
                        {' '}
                        {user.quantity}
                      </div>
                    </div>
                  ))}
                </ul>
              </div>
            </div>
          </div>
          <div className="comment-section">
            <Comment histComments={postDetails.comments} />
          </div>
          <div className="post-detail-tags">
            <div className="tags-label">Tags</div>
            <div className="tags-container">
              {postDetails.tags.map((tag) => (
                <button type="button" className="tag" key={tag}>{tag}</button>
              ))}
            </div>
          </div>
          <div className="buttons-container">
            <div className="desired-quantity">
              <input
                placeholder="desired quantity"
                onChange={(e) => setDesiredQuantity(e.target.value)}
              />
            </div>
            <div className="horizontal-buttons">
              <button
                disabled={!validQuantity}
                className="create-button"
                type="button"
                onClick={joinGroup}
              >
                Join
              </button>
              <button className="cancel-button" type="button" onClick={back}>
                Back
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  }
  return null;
}

export default PostDetails;
