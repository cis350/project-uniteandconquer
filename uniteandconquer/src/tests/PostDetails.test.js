/**
 * @jest-environment jsdom
 */

import React from 'react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import {
  render, screen, fireEvent,
} from '@testing-library/react';
import PostDetails from '../components/PostDetails';

// https://stackoverflow.com/questions/66284286/react-jest-mock-usenavigate

const mockedUsedLink = jest.fn();
const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  Link: () => mockedUsedLink,
  useNavigate: () => mockedUsedNavigate,
}));

test('post details title header', () => {
  render(<PostDetails />);
  const titleElement = screen.getByText('Item Details');
  expect(titleElement).toBeInTheDocument();
});

// test('go back and not join group', () => {
//   render(<PostDetails />);
//   fireEvent.click(screen.getByText('Back'));
// });

// test('inputting desired quantity before joining', () => {
//   render(<PostDetails />);
//   const input = screen.getByTestId('quantity-input');
//   fireEvent.change(input, { target: { value: '2' } });
//   expect(input.value).toBe('2');
// });

// test('inputting invalid format (string) desired quantity before joining', () => {
//   render(<PostDetails />);
//   const input = screen.getByTestId('quantity-input');
//   fireEvent.change(input, { target: { value: 'five' } });
//   expect(input.value).toBe('five');
// });

// test('join group', () => {
//   render(<PostDetails />);
//   const input = screen.getByTestId('quantity-input');
//   fireEvent.change(input, { target: { value: '2' } });
//   expect(input.value).toBe('2');
//   fireEvent.click(screen.getByText('Join'));
// });

// test('going back to home page', () => {
//   render(<PostDetails />);
//   fireEvent.click(screen.getByText('Back'));
// });

test('snapshot post-details', () => {
  const component = renderer.create(<PostDetails />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
