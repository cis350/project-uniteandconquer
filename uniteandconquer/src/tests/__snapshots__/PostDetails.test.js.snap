// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot post-details 1`] = `null`;
